### PR TITLE
SALTO-1624-stripe-improvements

### PR DESCRIPTION
**Pull request by tamtamirr**
Made Stripe's element ids more readable
---
Stripe's default id isn't very clear to human eye. Enriched most of the element ids with more fields to make them more readable.
___
_Release Notes_: 
stripe-adapter:
- Enriched element ids and filenames with constant fields.
- Plans are  no longer fetched. (Deprecated API)
- Prices are fetched per product and stored as a list under the field `product_prices` of each product.
---
_User Notifications_: 
-  Major changes on outdated Stripe records.

### CODE REVIEW

**Comment by tamtamirr** [comment]
You can see a sample updated workspace [here](https://github.com/tamtamirr/remote/tree/stripe-records-structure)

File: packages/stripe-adapter/test/adapter.test.ts [hunk 1]
```diff
@@ -14,255 +14,200 @@
...
-        ),
-        elementsSource: buildElementsSourceFromElements([]),
-      }).fetch({ progressReporter: { reportProgress: () => null } })
-      expect(elements.filter(isInstanceElement)).toHaveLength(17)
-
-      const countrySpec = elements.filter(isInstanceElement).find(e => e.elemID.getFullName() === 'stripe.country_spec.instance.AT')
-      expect(countrySpec).toBeDefined()
-      expect(countrySpec?.value).toEqual(expect.objectContaining({ id: 'AT' }))
-
+  describe('fetch', () => {
```
Comment (netama) [inline packages/stripe-adapter/test/adapter.test.ts:97]: can you please explain the goal of the refactor, and what was actually added in the tests?

**Review by netama — COMMENTED** [review]
added a few questions. since you mentioned you already went over this with @nitzan-haim and these changes are contained within the Stripe adapter, no blocker from my side if Nitzan approves it

File: packages/stripe-adapter/jest.config.js [hunk 2]
```diff
@@ -29,12 +29,13 @@ module.exports = deepMerge(
     : undefined,
     coverageThreshold: {
       'global': {
-        branches: 89,
-        functions: 94,
-        lines: 95,
-        statements: 95,
+        branches: 83,
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/jest.config.js:32]: the test is there, its odd that the coverage went down though. I'll inspect that

File: packages/stripe-adapter/e2e_test/adapter.test.ts [hunk 3]
```diff
@@ -37,49 +45,41 @@ describe('Stripe adapter E2E with real swagger and mock replies', () => {
...
-          'stripe.tax_rate_metadata',
-          'stripe.webhook_endpoint_metadata',
-        ]))
-      })
-    })
+  describe('fetches swagger types', () => {
+    it.each(EXPECTED_TYPES)(
+      '%s',
+      async expectedType => {
+        expect(fetchedSwaggerTypes).toContain(expectedType)
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/e2e_test/adapter.test.ts:81]: - Added a test that validates no additional types are fetched.
- Is it really gonna be drastically less efficient for bigger N? The business logic happens once in the beforeAll scope, we just take advantage of Jest's `each` for more specific failures.

In general, I think we should stick to tests that have more indicative error messages. I've had the experience of debugging those tests when changing Stripe's fetch config and they broke with less indicative reason.

File: packages/stripe-adapter/package.json [hunk 4]
```diff
@@ -58,6 +58,7 @@
     "eslint-plugin-react": "^7.14.3",
     "eslint-plugin-react-hooks": "^1.7.0",
     "jest": "^26.6.3",
+    "jest-extended": "^1.2.0",
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/package.json:61]: `jest-extended` shouldn't be blocked by `ts-jest`, it's a javascript package

File: packages/stripe-adapter/src/config.ts [hunk 5]
```diff
@@ -42,48 +44,82 @@ export type StripeConfig = {
...
   'coupons',
-  'plans',
-  'prices',
   'products',
   'reporting__report_types',
   'tax_rates',
   'webhook_endpoints',
 ]
 
+// noinspection UnnecessaryLocalVariableJS
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/src/config.ts:120]: Yes, it gives inspection (at least using `Webstorm`) that DEFAULT_INCLUDE_TYPES is redundant, but we already agreed it's better to keep it that way

File: packages/stripe-adapter/test/adapter.test.ts (hunk 1 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:97]: - Added a TestFactory that creates a test per included type, to make sure all of the types are being fetched. In contrast to the previous tests, this one **will break** when adding new include types and not handling them correctly.
- In order to keep the strict behaviour of the previous test, added a test that compares two sets of full element ids from stripe (In a way, it's even more strict since it works on bigger sets of data and not just a single element per type).

The tests above replace the concept of the previous test that validated specific elements of each type were fetched.

File: packages/stripe-adapter/test/instances_full_names.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
...iation_itemized_1@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_2@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_3@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_4@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_5@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_summary_1@uvv', 'stripe.reporting_report_type.instance.payouts_itemized_1@v', 'stripe.reporting_report_type.instance.payouts_itemized_2@v', 'stripe.reporting_report_type.instance.payouts_itemized_3@v', 'stripe.reporting_report_type.instance.payouts_summary_1@v', 'stripe.reporting_report_type.instance.tax_product_filing_itemized_1@uvvv', 'stripe.reporting_report_type.instance.tax_product_transactions_itemized_1@uvvv', 'stripe.reporting_report_type.instance.tax_product_transactions_summarized_1@uvvv', 'stripe.tax_rate.instance.Sales_tax_txr_1Jwt1JHipyrr1EYi3coTN5Y9_AL_3@suuuu', 'stripe.tax_rate.instance.Sales_tax_txr_1JtCkIHipyrr1EYiy5LweTs6_AL_3@suuuu', 'stripe.tax_rate.instance.Sales_tax_txr_1IcrqNHipyrr1EYiTpN0iuVj_AL_88@suuuu', 'stripe.webhook_endpoint.instance.we_1IcrkHHipyrr1EYiSGhlscV7']
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/test/instances_full_names.ts:16]: - Figured because it's a big set of data reformatting this will take many lines, but it seems like it's not affected by lint at all, will remove this file.
- Is it bad to include all of the examples and be more strict?

File: packages/stripe-adapter/jest.config.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/jest.config.js]: Oddly enough, this Saturday the coverage went way up 😂  I'll make sure everything passes on circle CI

File: packages/stripe-adapter/src/config.ts [hunk 7]
```diff
@@ -42,48 +44,82 @@ export type StripeConfig = {
...
-    { originalName: 'v1__coupons', newName: 'coupons' },
-    { originalName: 'v1__plans', newName: 'plans' },
-    { originalName: 'v1__prices', newName: 'prices' },
-    { originalName: 'v1__products', newName: 'products' },
-    { originalName: 'v1__reporting__report_types', newName: 'reporting__report_types' },
-    { originalName: 'v1__tax_rates', newName: 'tax_rates' },
-    { originalName: 'v1__webhook_endpoints', newName: 'webhook_endpoints' },
-  ],
-  additionalTypes: [],
+export const DEFAULT_API_DEFINITIONS: StripeApiConfig = {
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/src/config.ts:89]: No hidden changes, as a fresh newbe a couple weeks ago made it match the newer adapter config structure (`jira-adapter`). Could revert if needed

File: packages/stripe-adapter/e2e_test/adapter.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/e2e_test/adapter.test.ts:78]: UPDATE
--------
- The original test did check for contains and not pure equals. Removed the test I've mentioned from my previous comment.
- Removed the EXPECTED_TYPES constant.

File: packages/stripe-adapter/test/mock_replies.json [hunk 8]
```diff
@@ -1,180 +1,13 @@
...
+]
```
Comment (nitzan-haim) [inline packages/stripe-adapter/test/mock_replies.json:13]: can you please specify what were the changes in this file? it's hard to follow. Did you re-run the script?

File: packages/stripe-adapter/test/adapter.test.ts [hunk 9]
```diff
@@ -14,255 +14,201 @@
...
...iation_itemized_1@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_2@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_3@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_4@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_itemized_5@uvv', 'stripe.reporting_report_type.instance.payout_reconciliation_summary_1@uvv', 'stripe.reporting_report_type.instance.payouts_itemized_1@v', 'stripe.reporting_report_type.instance.payouts_itemized_2@v', 'stripe.reporting_report_type.instance.payouts_itemized_3@v', 'stripe.reporting_report_type.instance.payouts_summary_1@v', 'stripe.reporting_report_type.instance.tax_product_filing_itemized_1@uvvv', 'stripe.reporting_report_type.instance.tax_product_transactions_itemized_1@uvvv', 'stripe.reporting_report_type.instance.tax_product_transactions_summarized_1@uvvv', 'stripe.tax_rate.instance.Sales_tax_txr_1Jwt1JHipyrr1EYi3coTN5Y9_AL_3@suuuu', 'stripe.tax_rate.instance.Sales_tax_txr_1JtCkIHipyrr1EYiy5LweTs6_AL_3@suuuu', 'stripe.tax_rate.instance.Sales_tax_txr_1IcrqNHipyrr1EYiTpN0iuVj_AL_88@suuuu', 'stripe.webhook_endpoint.instance.we_1IcrkHHipyrr1EYiSGhlscV7']
```
Comment (nitzan-haim) [inline packages/stripe-adapter/test/adapter.test.ts:61]: since it's an open-source project, make sure there is no hardcoded personal information (also in the mock_replies.json. I see your name in several places, don't know if you're ok with that

File: packages/stripe-adapter/test/adapter.test.ts [hunk 10]
```diff
@@ -14,255 +14,201 @@
...
-        ),
-        config: new InstanceElement(
+      it('fetches instances of modified type: "product"', async () => {
+        const config = new InstanceElement(
           'config',
           configType,
           {
             [FETCH_CONFIG]: {
-              includeTypes: ['prices'],
+              includeTypes: DEFAULT_INCLUDE_TYPES,
```
Comment (nitzan-haim) [inline packages/stripe-adapter/test/adapter.test.ts:185]: The point of the previous test ("should fetch only the relevant types") was to test the `includeTypes` behavior. Please add a test that makes sure that the fetched types are only the ones included in `includeTypes` (or point out one that does that that I might have missed)

File: packages/stripe-adapter/test/adapter.test.ts [hunk 11]
```diff
@@ -14,255 +14,201 @@
...
-        elementsSource: buildElementsSourceFromElements([]),
+        )
+        const fetchedTypes = (await fetchInstances(config)).map(i => i.elemID.typeName)
+        expect(fetchedTypes).toContain('product')
       })
-      await expect(operations.deploy({ changeGroup: { groupID: '', changes: [] } })).rejects.toThrow(new Error('Not implemented.'))
     })
   })
+
+  it('throws Error on deploy', async () => {
```
Comment (nitzan-haim) [inline packages/stripe-adapter/test/adapter.test.ts:205]: missing `describe('deploy')`

File: packages/stripe-adapter/jest.config.js [hunk 12]
```diff
@@ -29,12 +29,13 @@ module.exports = deepMerge(
...
     coverageThreshold: {
       'global': {
-        branches: 89,
-        functions: 94,
-        lines: 95,
-        statements: 95,
+        branches: 100,
+        functions: 95,
+        lines: 99,
+        statements: 99,
```
Comment (nitzan-haim) [inline packages/stripe-adapter/jest.config.js:35]: not sure we want to be that strict with the coverage requirements, @netama WDYT?

**Review by nitzan-haim — COMMENTED** [review]
Added some comments :)

File: packages/stripe-adapter/test/mock_replies.json (hunk 8 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/mock_replies.json:13]: Yes, ran the script for the new config

File: packages/stripe-adapter/test/adapter.test.ts (hunk 9 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:61]: I'm cool with that :) I mean, we got our name on gitblame after all 😈

File: packages/stripe-adapter/test/adapter.test.ts (hunk 10 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:186]: It's there :)
Refer to `custom config`: `fetches included types`

File: packages/stripe-adapter/test/adapter.test.ts (hunk 11 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts]: v

File: packages/stripe-adapter/jest.config.js (hunk 12 above)
Comment (tamtamirr) [reply packages/stripe-adapter/jest.config.js:35]: After all, a coverage is a tool for us to keep track on how well we cover our tests. If the coverage will go down for valid reasons, the coverage will change accordingly

File: packages/stripe-adapter/test/adapter.test.ts (hunk 10 above)
Comment (nitzan-haim) [reply packages/stripe-adapter/test/adapter.test.ts:186]: @tamtamirr can you put a link to the test? I'm not sure I understand

File: packages/stripe-adapter/test/mock_replies.json (hunk 8 above)
Comment (nitzan-haim) [reply packages/stripe-adapter/test/mock_replies.json:13]: @tamtamirr alright, and just making sure - there is at least one instance of each type? and all 6 types that we support are included?

File: packages/stripe-adapter/test/adapter.test.ts [hunk 13]
```diff
@@ -14,255 +14,204 @@
...
-  const actual = jest.requireActual('@salto-io/adapter-components')
-  return {
-    ...actual,
-    elements: {
-      ...actual.elements,
-      swagger: {
-        ...actual.elements.swagger,
-        generateTypes: jest.fn().mockImplementation(generateMockTypes),
+  const SINGULAR_TYPES = ['country_spec', 'coupon', 'product', 'reporting_report_type', 'tax_rate', 'webhook_endpoint']
+  const pluralToSingularTypes = _.zipObject(DEFAULT_INCLUDE_TYPES, SINGULAR_TYPES)
```
Comment (amitgoldarb) [inline packages/stripe-adapter/test/adapter.test.ts:44]: This doesn't look the same as in the previous version. Is that on purpose?

File: packages/stripe-adapter/test/adapter.test.ts [hunk 14]
```diff
@@ -14,255 +14,204 @@
...
+            }
+          )
+          const instances = await fetchInstances(config)
+          fetchedInstancesTypes = new Set(instances.map(e => e.elemID.typeName))
+        })
+
+        it.each(SINGULAR_INCLUDE_TYPES)(
+          '%s',
+          includedType => {
+            expect(fetchedInstancesTypes).toContain(includedType)
```
Comment (amitgoldarb) [inline packages/stripe-adapter/test/adapter.test.ts:170]: We had more validations in the previous version didn't we? We validated that each instance included an ID and we also verified specifically which id it was.

File: packages/stripe-adapter/test/adapter.test.ts [hunk 15]
```diff
@@ -14,255 +14,204 @@
...
+        )
 
-      const webhookEndpoint = elements.filter(isInstanceElement).find(e => e.elemID.getFullName() === 'stripe.webhook_endpoint.instance.we_1IcrkHHipyrr1EYiSGhlscV7')
-      expect(webhookEndpoint).toBeDefined()
-      expect(webhookEndpoint?.value).toEqual(expect.objectContaining({ id: 'we_1IcrkHHipyrr1EYiSGhlscV7' }))
-    })
-  })
+        it('doesn\'t fetch additional types', () => {
+          const additionalTypes: string[] = Array.from(fetchedInstancesTypes).filter(notIncluded)
+          expect(additionalTypes).toBeEmpty()
```
Comment (amitgoldarb) [inline packages/stripe-adapter/test/adapter.test.ts:176]: I don't think we need so many variables for this test. It makes it harder to follow on the eyes:
```suggestion
          expect(Array.from(fetchedInstancesTypes).filter(type => !SINGULAR_INCLUDE_TYPES.includes(type))).toHaveLength(0)
```

File: packages/stripe-adapter/test/adapter.test.ts [hunk 16]
```diff
@@ -14,255 +14,204 @@
...
-      expect(coupon).toBeDefined()
-      expect(coupon?.value).toEqual(expect.objectContaining({ id: '2JI7xnRz' }))
+    beforeAll(async () => {
+      fetchedInstances = await fetchInstances()
+    })
 
+    it('fetches all instances', () => {
+      const fetchedInstancesFullNames = fetchedInstances
+        .map(instance => instance.elemID.getFullName())
+      expect(fetchedInstancesFullNames).toIncludeSameMembers(EXPECTED_INSTANCES_FULL_NAMES)
```
Comment (amitgoldarb) [inline packages/stripe-adapter/test/adapter.test.ts:109]: Can we sort and compare?

File: packages/stripe-adapter/test/adapter.test.ts (hunk 10 above)
Comment (amitgoldarb) [reply packages/stripe-adapter/test/adapter.test.ts:186]: I didn't follow either TBH.

**Review by amitgoldarb — COMMENTED** [review]
Great job!
Added some comments let's discuss f2f

File: packages/stripe-adapter/jest.config.js (hunk 12 above)
Comment (netama) [reply packages/stripe-adapter/jest.config.js:35]: I agree with @nitzan-haim , the coverage thresholds are an indication of what we think is "good enough". I don't think we should rqeuire coverage above ~95%, @ori-moisis wdyt?

Comment (amitgoldarb) [reply packages/stripe-adapter/jest.config.js:35]: @netama This is different than how we work in the SAAS. In the SAAS we generally raise and lower our coverage quite fluildy based on our PR. If we lower we should have an explanataion (there are good reasons). If we don't then  we shouldn't lower it.

Comment (ori-moisis) [reply packages/stripe-adapter/jest.config.js:35]: I don't think we should normalize changing the target thresholds (we know if a PR lowers coverage with coveralls, we don't need to rely on the thresholds), in the past this was a point of contention.
I won't stop you from making the threshold too high though, I don't mind having the standard higher :)
(just know that there is a very small number of acceptable reasons to lower thresholds in future PRs)

File: packages/stripe-adapter/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/stripe-adapter/package.json:61]: I like this feature especially toSatisfyAll, I think the output was more readable. Nice!

File: packages/stripe-adapter/src/config.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/stripe-adapter/src/config.ts:120]: Discussed f2f. Agreed this is best for now

File: packages/stripe-adapter/test/adapter.test.ts (hunk 10 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:186]: https://github.com/tamtamirr/salto/blob/a4cfb7390b7f8daa5288ee2520c95e94b840209b/packages/stripe-adapter/test/adapter.test.ts#L143

File: packages/stripe-adapter/e2e_test/adapter.test.ts [hunk 17]
```diff
@@ -17,69 +17,66 @@ import { isObjectType } from '@salto-io/adapter-api'
...
   beforeAll(async () => {
     credLease = await credsLease()
     const adapterAttr = realAdapter({ credentials: credLease.value })
-    adapter = adapterAttr.adapter
+    const { elements } = await adapterAttr.adapter.fetch({
+      progressReporter:
+        { reportProgress: () => null },
+    })
+    fetchedSwaggerTypes = elements
+      .filter(isObjectType)
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/e2e_test/adapter.test.ts:38]: v

File: packages/stripe-adapter/e2e_test/adapter.test.ts [hunk 18]
```diff
@@ -17,69 +17,66 @@ import { isObjectType } from '@salto-io/adapter-api'
...
-          'stripe.plan_tier',
-          'stripe.transform_usage',
-          'stripe.tax_rate_metadata',
-          'stripe.webhook_endpoint_metadata',
-        ]))
-      })
-    })
+  describe('fetches swagger types', () => {
+    it.each([
+      'stripe.country_specs',
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/e2e_test/adapter.test.ts:49]: Checked. All of the types were plan related. However, we should still fetch everything but stripe.plans. Added the rest, removed `stripe.plan`

File: packages/stripe-adapter/test/adapter.test.ts (hunk 14 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:170]: This test should be more accurate and will break when not handling one of the included types correctly (The previous test didn't).

Separated the hardcore logic to [this](https://github.com/tamtamirr/salto/blob/a4cfb7390b7f8daa5288ee2520c95e94b840209b/packages/stripe-adapter/test/adapter.test.ts#L106) test that strictly checks for full element id names

File: packages/stripe-adapter/test/adapter.test.ts (hunk 15 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:176]: In general, I try to keep my expect().matcher() lines short, especially for error reporting and stacktrace.
However I do agree it's not fun to jump around the code for a quite simple filter, so moved the `notIncluded` into the test.

File: packages/stripe-adapter/test/adapter.test.ts (hunk 16 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:109]: discussed

File: packages/stripe-adapter/test/adapter.test.ts (hunk 13 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:44]: Yes, that way you could add more inputs and modify the INCLUDE_TYPE to any supported types so the test won't break. Hardcoding them would cause a bigger refactor on change.

File: packages/stripe-adapter/e2e_test/adapter.test.ts [hunk 19]
```diff
@@ -17,69 +17,66 @@ import { isObjectType } from '@salto-io/adapter-api'
...
     }
   })
-
-  describe('fetch', () => {
-    describe('swagger types only', () => {
-      it('should generate the right type elements on fetch', async () => {
-        const { elements } = await adapter.fetch({ progressReporter:
-          { reportProgress: () => null } })
-
-        expect(elements.every(isObjectType)).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (tamtamirr) [reply packages/stripe-adapter/e2e_test/adapter.test.ts:46]: v, added test

File: packages/stripe-adapter/test/adapter.test.ts (hunk 9 above)
Comment (tamtamirr) [reply packages/stripe-adapter/test/adapter.test.ts:63]: v, we should discuss IRL though whats the right way to work on the swagger adapters. The adapter tests force to use @netama 's script after every change in the conifg, and the process of prettifying the output can be long and excessive.

**Review by amitgoldarb — APPROVED** [review]
LGTM!
Great work
